Applications/Spire/Source/Ui/DropDownList.cpp

```diff
@@ -0,0 +1,247 @@
+#include "Spire/Ui/DropDownList.hpp"
+#include <QFocusEvent>
+#include <QKeyEvent>
+#include <QScrollBar>
+#include <QVBoxLayout>
+#include "Spire/Spire/Dimensions.hpp"
+#include "Spire/Ui/ScrollArea.hpp"
+
+using namespace boost::signals2;
+using namespace Spire;
+
+namespace {
+  const auto BORDER_PADDING = 2;
+}
+
+DropDownList::DropDownList(std::vector<DropDownItem*> items,
+    bool is_click_activated, QWidget* parent)
+    : DropDownWindow(is_click_activated, parent),
+      m_max_displayed_items(5) {
+  m_scroll_area = new ScrollArea(this);
+  m_scroll_area->setFocusProxy(parent);
+  m_scroll_area->setWidgetResizable(true);
+  auto main_widget = new QWidget(this);
+  m_layout = new QVBoxLayout(main_widget);
+  m_layout->setContentsMargins({});
+  m_layout->setSpacing(0);
+  m_scroll_area->setWidget(main_widget);
+  set_widget(m_scroll_area);
+  set_items(items);
+  parent->installEventFilter(this);
+}
+
+bool DropDownList::eventFilter(QObject* watched, QEvent* event) {
+  if(watched == parent()) {
+    if(event->type() == QEvent::KeyPress) {
+      auto e = static_cast<QKeyEvent*>(event);
+      if(e->key() == Qt::Key_Up) {
```

Comment:
Use `switch` with fallthrough for the last case.

Reply:
By fallthrough do you mean omit the default case?

Reply:
Fallthrough means absense of `break` statement at the end of a case. You can structure this control flow in the following way:
```
case A:
  // action1
  break;
case B:
  // action2
  break;
case C:
case D:
  // action3
  break;
```